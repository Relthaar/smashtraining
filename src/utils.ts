/**
 * Return the key-value pairs in the object, but preserve detailed type
 * information about the keys. That is, if the set of keys is statically-known,
 * then the returned entries will preserve that type.
 */
export function entries<T extends string, U>(
  obj: Partial<Record<T, U>>,
): Array<[T, U]> {
  const result: Array<[T, U]> = [];
  for (const [k, v] of Object.entries(obj)) {
    if (v !== undefined) {
      result.push([k as T, v as U]);
    }
  }
  return result;
}

export function upcast<From extends To, To>(t: From): To {
  return t;
}

export async function log(message: string, ...params: any[]): Promise<void> {
  // tslint:disable-next-line: no-console
  console.log(message, ...params);
}

export function assert(condition: boolean, message: string): condition is true {
  if (!condition) {
    throw new Error(`Assertion failed: ${message}`);
  }
  return condition;
}

export function unreachable(arg: never, message: string): never {
  throw new Error(`unreachable: ${message}`);
}

/**
 * Determine if two objects are equal.
 *
 * This code is from `vuetify/src/util/helpers.ts` (Vuetify license: MIT), but
 * inlined here because configuring Jest to import from TypeScript files in
 * `node_modules` seems to be difficult.
 */
export function deepEqual(a: any, b: any): boolean {
  if (a === b) {
    return true;
  }

  if (a instanceof Date && b instanceof Date) {
    // If the values are Date, they were convert to timestamp with getTime and compare it
    if (a.getTime() !== b.getTime()) {
      return false;
    }
  }

  if (a !== Object(a) || b !== Object(b)) {
    // If the values aren't objects, they were already checked for equality
    return false;
  }

  const props = Object.keys(a);

  if (props.length !== Object.keys(b).length) {
    // Different number of props, don't bother to check
    return false;
  }

  return props.every(p => deepEqual(a[p], b[p]));
}
