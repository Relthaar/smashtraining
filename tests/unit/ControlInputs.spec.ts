import ControlInputs from "@/components/ControlInputs.vue";
import { shallowMount } from "@vue/test-utils";
import Vue from "vue";
import Vuetify from "vuetify";
import { flagConsoleErrors, ignoreConsoleErrors } from "../utils";

describe("HelloWorld.vue", () => {
  beforeAll(() => {
    Vue.use(Vuetify);
  });

  const token = flagConsoleErrors();
  beforeEach(token.beforeEach);
  afterEach(token.afterEach);

  it("renders inputs", () => {
    const inputs = "u d l r . U D L R - a + b sh fh";
    const wrapper = shallowMount(ControlInputs, {
      propsData: { inputs },
    });
    expect(wrapper.text()).toMatchSnapshot();
  });

  it("rejects unknown inputs", () => {
    ignoreConsoleErrors(() => {
      const inputs = "foo";
      expect(() =>
        shallowMount(ControlInputs, {
          propsData: { inputs },
        }),
      ).toThrowError("Unknown control input type: foo");
    });
  });
});
