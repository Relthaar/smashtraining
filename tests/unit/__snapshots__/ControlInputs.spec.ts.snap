// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HelloWorld.vue renders inputs 1`] = `"Gently tap up Gently tap down Gently tilt backward Gently tilt forward Release controls Hard tap up Hard tap down Hard tap backward Hard tap forward Slight pause Press A Simultaneous input Press B Short-hop Full hop"`;
